### PR TITLE
Enable pytorch 1.5.

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
pytorch 1.5 came out and broke our tests. This turns out to because our method for hogwild no longer works. Unfortunately, I think we're going to have to simply sunset hogwild, but this requires a lot of doc updates etc. The best way is to do this before 1.6, and for now just tell users that hogwild is 1.4 or earlier.

So in summary:
- Make the helper decorators for pytorch a little more usable.
- Upgrade the installation caches for CI
- Disable hogwild only when in 1.5.

**Testing steps**
Local testing, but CI requires a lot of updates.

### CODE REVIEW

File: parlai/utils/testing.py [hunk 1]
```diff
@@ -89,15 +89,31 @@ def skipIfCircleCI(testfn, reason='Test disabled in CircleCI'):
...
         skip = True
     else:
-        from packaging import version
+        import packaging.version
 
-        skip = version.parse(torch.__version__) < version.parse('1.4.0')
-    return unittest.skipIf(skip, reason)(testfn)
+        target = packaging.version.parse(version)
+        skip = packaging.version.parse(torch.__version__) < target
+    reason = 'Test requires pytorch {version} or newer'
```
Comment (klshuster) [inline parlai/utils/testing.py:101]: nit: prob need an `f""` string here

File: parlai/utils/testing.py [hunk 2]
```diff
@@ -89,15 +89,31 @@ def skipIfCircleCI(testfn, reason='Test disabled in CircleCI'):
...
+def skipIfTorchVersion(version):
+    skip = False
+    if not TORCH_AVAILABLE:
+        skip = True
+    else:
+        import packaging.version
+
+        target = packaging.version.parse(version)
+        skip = packaging.version.parse(torch.__version__) >= target
+    reason = 'Test requires pytorch OLDER than {version}'
```
Comment (klshuster) [inline parlai/utils/testing.py:114]: ditto

File: tests/test_memnn.py [hunk 3]
```diff
@@ -73,6 +74,7 @@ def test_labelcands_multi(self):
         self.assertGreater(valid['hits@1'], 0.95)
         self.assertGreater(test['hits@1'], 0.95)
 
+    @unittest.skip
```
Comment (klshuster) [inline tests/test_memnn.py:77]: should we just delete the test if we're skipping it?

File: .circleci/config.yml [hunk 4]
```diff
@@ -147,12 +146,12 @@ jobs:
...
       - restore_cache:
-          key: deps-v5-osx-{{ checksum "requirements.txt" }}
+          key: deps-v5-osx-testing1-{{ checksum "requirements.txt" }}
       - <<: *setup
-      - <<: *installdeps
       - <<: *installtorchcpuosx
+      - <<: *installdeps
       - save_cache:
-          key: deps-v5-osx-{{ checksum "requirements.txt" }}
+          key: deps-v5-osx-testing1-{{ checksum "requirements.txt" }}
```
Comment (klshuster) [inline .circleci/config.yml:154]: just curious why you added the `testing1` for all the cache keys

Comment (stephenroller) [reply .circleci/config.yml:154]: Bc I was testing, of course

File: tests/test_memnn.py (hunk 3 above)
Comment (stephenroller) [reply tests/test_memnn.py:77]: whoops!

**Comment by stephenroller** [comment]
Sigh, it looks like pytorch 1.5 hard breaks model parallelism (pytorch/pytorch#38605).

**Comment by stephenroller** [comment]
We'll revive for 1.5.1.

**Pull request closed without merging**